### PR TITLE
satellite/metainfo: add buckets RPC and database

### PR DESCRIPTION
**Pull request by JessicaGreben**
What/Why:
This PR is one part of the [ticket](https://storjlabs.atlassian.net/browse/V3-1967) to move bucket metadata into the satellite masterDB. I'm breaking this ticket into many small PRs so they don't get too big.
We want to move bucket metadata out of the pointerDB since its been a source of bugs and more complex code treating buckets as objects. This PR updates the satellite metainfo service to use the bucket RPCs. Changes include:
- add metainfo endpoints to interact for bucket RPCs
- update metainfo service with bucket methods
- add a bucket database interface and satellitedb bucket methods to interact with buckets in masterDB
- add filtering for bucket listing results on Satellite side by macaroon restrictions

Please describe the tests:
- Added tests for the bucketDB code.
 
Please describe the performance impact:
- Overall these changes should improve performance. We will write some benchmarks once the uplink portion of this is implemented

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/buckets/db.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package buckets
```
Comment (egonelbre) [inline satellite/buckets/db.go]: Having this all in metainfo is fine since it's a single endpoint.

File: pkg/storj/bucket.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+// Bucket contains information about a specific bucket
+type Bucket struct {
+	ID                   uuid.UUID
+	Name                 string
+	ProjectID            uuid.UUID
+	Attribution          string
+	Created              time.Time
+	PathCipher           CipherSuite
+	SegmentsSize         int64
```
Comment (egonelbre) [inline pkg/storj/bucket.go]: These are defaults not the sizes.

File: pkg/macaroon/apikey.go [hunk 3]
```diff
@@ -131,6 +131,27 @@ func (a *APIKey) Check(ctx context.Context, secret []byte, action Action, revoke
 	return nil
 }
 
+// GetAllowedBuckets returns a list of all the allowed bucket paths that match the Action operation
+func (a *APIKey) GetAllowedBuckets(ctx context.Context, action Action) (allowedBuckets map[string]struct{}, err error) {
```
Comment (egonelbre) [inline pkg/macaroon/apikey.go]: @zeebo is there a better way to handle this?

File: satellite/metainfo/metainfo.go [hunk 4]
```diff
@@ -631,37 +632,193 @@ func (endpoint *Endpoint) ProjectInfo(ctx context.Context, req *pb.ProjectInfoRe
...
-func (endpoint *Endpoint) SetBucketAttribution(ctx context.Context, req *pb.BucketSetAttributionRequest) (_ *pb.BucketSetAttributionResponse, err error) {
-	defer mon.Task()(&ctx)(&err)
-	// TODO: placeholder to implement pb.MetainfoServer interface.
-	return &pb.BucketSetAttributionResponse{}, err
+func getAllowedBuckets(ctx context.Context, action macaroon.Action) (allowedBuckets map[string]struct{}, err error) {
+	keyData, ok := auth.GetAPIKey(ctx)
+	if !ok {
+		return nil, status.Errorf(codes.Unauthenticated, "Invalid API credential")
+	}
+	key, err := macaroon.ParseAPIKey(string(keyData))
```
Comment (egonelbre) [inline satellite/metainfo/metainfo.go]: Isn't this parsed already in keyinfo above?

File: satellite/metainfo/metainfo.go [hunk 5]
```diff
@@ -631,37 +632,193 @@ func (endpoint *Endpoint) ProjectInfo(ctx context.Context, req *pb.ProjectInfoRe
...
+	}
+
+	items, err := endpoint.metainfo.ListBuckets(ctx, keyInfo.ProjectID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "ListBuckets: %v", err)
+	}
+
+	allowedBuckets, err := getAllowedBuckets(ctx, action)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "getAllowedBuckets: %v", err)
```
Comment (egonelbre) [inline satellite/metainfo/metainfo.go]: This is hiding the actual error from getAllowedBuckets

File: satellite/metainfo/metainfo.go [hunk 6]
```diff
@@ -631,37 +632,193 @@ func (endpoint *Endpoint) ProjectInfo(ctx context.Context, req *pb.ProjectInfoRe
...
+	keyInfo, err := endpoint.validateAuth(ctx, macaroon.Action{
+		Op:            macaroon.ActionDelete,
+		Bucket:        req.Name,
+		EncryptedPath: []byte{},
+		Time:          time.Now(),
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Unauthenticated, err.Error())
+	}
+
```
Comment (mniewrzal) [inline satellite/metainfo/metainfo.go]: maybe we should add bucket name validation here to avoid executing rest of logic?

File: satellite/metainfo/metainfo.go [hunk 7]
```diff
@@ -631,37 +632,193 @@ func (endpoint *Endpoint) ProjectInfo(ctx context.Context, req *pb.ProjectInfoRe
...
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+
+	err = endpoint.metainfo.CreateBucket(ctx, bucket)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	return &pb.BucketCreateResponse{}, nil
```
Comment (mniewrzal) [inline satellite/metainfo/metainfo.go]: this should return `Bucket` in response

File: satellite/metainfo/metainfo.go [hunk 8]
```diff
@@ -631,37 +632,193 @@ func (endpoint *Endpoint) ProjectInfo(ctx context.Context, req *pb.ProjectInfoRe
...
+			bucketItems[i] = &pb.BucketListItem{
+				Name:      []byte(item.Name),
+				CreatedAt: item.Created,
+			}
+		}
+	}
+
+	return &pb.BucketListResponse{
+		Items: bucketItems,
+		More:  false,
```
Comment (mniewrzal) [inline satellite/metainfo/metainfo.go]: do you plan to support `more` in this PR?

File: satellite/satellitedb/buckets.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+		if err != nil {
+			return buckets, err
+		}
+		buckets = append(buckets, item)
+	}
+	return buckets, err
+}
+
+func convertDBXtoBucket(dbxBucket *dbx.BucketMetainfo) (bucket storj.Bucket, err error) {
+	id, err := uuid.Parse(string(dbxBucket.Id))
```
Comment (mniewrzal) [inline satellite/satellitedb/buckets.go]: I think this should be done more like this https://github.com/storj/storj/blob/0e528bc56ea05d896b490b3146e704492f2e9c3b/cmd/satellite/reports/attribution.go#L98

File: satellite/metainfo/metainfo.go (hunk 8 above)
Comment (JessicaGreben) [reply satellite/metainfo/metainfo.go]: yes, in the description of this PR, it says that part is WIP

File: pkg/storj/bucket.go (hunk 2 above)
Comment (JessicaGreben) [reply pkg/storj/bucket.go]: I feel like this Bucket struct is used in other places where these are the actual sizes and not the defaults.  Or I guess what does "defaults" mean?  What is set by default in the code? Or what value gets passed into the uplink config? Or what?

File: satellite/metainfo/metainfo.go (hunk 8 above)
Comment (mniewrzal) [reply satellite/metainfo/metainfo.go]: ah, yes :) thanks

File: pkg/macaroon/apikey.go (hunk 3 above)
Comment (zeebo) [reply pkg/macaroon/apikey.go]: No, I don't see a better way to do this. Did you have any specific concerns?

File: satellite/metainfo/metainfo.go (hunk 4 above)
Comment (zeebo) [reply satellite/metainfo/metainfo.go]: No, but the `validateAuth` function could be extended to return the parsed key. Personally, I also think it's fine to just parse it twice.

File: pkg/storj/bucket.go (hunk 2 above)
Comment (mniewrzal) [reply pkg/storj/bucket.go]: buckets in general doesn't' have a size or redundancy scheme. these values are stored as defaults for objects uploads.

File: satellite/metainfo/metainfo.go (hunk 5 above)
Comment (JessicaGreben) [reply satellite/metainfo/metainfo.go]: how does it hide the error?

File: pkg/macaroon/apikey.go (hunk 3 above)
Comment (egonelbre) [reply pkg/macaroon/apikey.go]: Just seemed something that could look nicer.

File: satellite/metainfo/metainfo.go (hunk 5 above)
Comment (egonelbre) [reply satellite/metainfo/metainfo.go]: getAllowedBuckets internally returns `codes.Unauthenticated`, which is the thing that should be sent to the client, here it's overriding the error code with `codes.Internal`.

File: satellite/metainfo/db_test.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+var testBucketNames = []string{"aaa", "bbb", "mmm", "qqq", "zzz",
+	"test.bucket", "123", "0test", "999", "test-bucket.thing",
+}
+
+func setup(ctx context.Context, bucketsDB metainfo.BucketsDB, projectID uuid.UUID) {
+	for _, bucket := range testBucketNames {
+		b := setupBucket(bucket, projectID)
+		_, err := bucketsDB.CreateBucket(ctx, b)
+		if err != nil {
+			fmt.Println(err)
```
Comment (egonelbre) [inline satellite/metainfo/db_test.go]: needs to be checked

File: satellite/metainfo/db_test.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+		}
+	}
+}
+
+func teardown(ctx context.Context, bucketsDB metainfo.BucketsDB, projectID uuid.UUID) {
+	for _, bucket := range testBucketNames {
+		err := bucketsDB.DeleteBucket(ctx, []byte(bucket), projectID)
+		if err != nil {
+			fmt.Println(err)
```
Comment (egonelbre) [inline satellite/metainfo/db_test.go]: needs to be checked

File: pkg/pb/encryption.proto [hunk 12]
```diff
@@ -11,13 +11,9 @@ message EncryptionParameters {
...
-    ENC_AESGCM = 2;
-    ENC_SECRETBOX = 3;
-  }
-
-  Type type = 1;
+enum CipherSuite {
+  ENC_UNSPECIFIED = 0;
+  ENC_NULL = 1;
+  ENC_AESGCM = 2;
+  ENC_SECRETBOX = 3;
```
Comment (zeebo) [inline pkg/pb/encryption.proto]: This change seems unrelated and I don't understand it.

File: satellite/satellitedb/buckets.go [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+	return bucket, nil
+}
+
+// GetBucket returns a bucket
+func (db *bucketsDB) GetBucket(ctx context.Context, bucketName []byte, projectID uuid.UUID) (bucket storj.Bucket, err error) {
+	defer mon.Task()(&ctx)(&err)
+	dbxBucket, err := db.db.Get_BucketMetainfo_By_ProjectId_And_Name(ctx,
+		dbx.BucketMetainfo_ProjectId(projectID[:]),
+		dbx.BucketMetainfo_Name(bucketName),
+	)
```
Comment (zeebo) [inline satellite/satellitedb/buckets.go]: unchecked err

File: pkg/pb/encryption.proto (hunk 12 above)
Comment (JessicaGreben) [reply pkg/pb/encryption.proto]: re: https://github.com/storj/storj/pull/2383#discussion_r299377712
